clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -461,6 +463,12 @@ private static int parseAcks(String acksString) {
                         " specified in value.serializer");
             }
 
+            if(checkExpectedOffsets && record.expectedOffset() == -1L) {
```

Comment:
if we'd been working on this patch a year ago, this would have been way easier (for us).  The offset field in each message was literally unused; so we could have had the server validate if the offset value was non-zero